design/ErrorHandling.md

```diff
@@ -0,0 +1,160 @@
+# Error Handling
+
+The simplest design for error handling would be to do it synchronously, for example with Javascript exceptions and object creation returning `null`.
+However, this would introduce a lot of synchronization points for multi-threaded/multi-process WebGPU implementations, making too slow to be useful.
+
+There are a number of cases that developers or applications need error handling for:
+
+ - *Debugging*: Getting errors synchronously during development, to break in to the debugger.
+ - *Telemetry*: Collecting error logs in deployment, for bug reporting and telemetry.
+ - *Recovery*: Recovering from recoverable errors (like out-of-memory on resource creation).
+ - *Fallback*: Tearing down the application and falling back, e.g. to WebGL, 2D Canvas, or static content.
+
+Meanwhile, error handling should not make the API clunky to use.
+
+There are several types of WebGPU calls that get their error handled differently:
+
+ - *Creation*: WebGPU Object creation.
+ - *Encoding*: Recording of GPU commands in `WebGPUCommandEncoder`.
+ - *Operations*: Other commands, such as `WebGPUQueue.submit`.
+
+## *Telemetry*: Error Logging
+
+`WebGPUDevice` keeps a log of errors that happened on each WebWorker which the application can query asynchronously.
+These errors are not directly associated with the objects or operations that produced them.
+They should not be used by applications to recover from expected errors.
+Instead, they are used for handling unexpected errors in deployment, for bug reporting and telemetry.
+
+```
+enum WebGPULogEntryType {
+    "device-lost",
+    "validation-error",
+    "recoverable-out-of-memory",
+};
+
+interface WebGPULogEntry {
+    readonly attribute WebGPULogEntryType type;
+    readonly attribute DOMString? reason;
+};
+
+partial interface WebGPUDevice {
+    Promise<sequence<WebGPULogEntry>> getCurrentErrorLog();
+};
+```
+
+`WebGPUObjectStatusType` makes a distinction between several error types:
+
+ - `"validation-error"`: validation of some WebGPU call failed - including if an argument was an "invalid object".
+   (Either the application did something wrong, or it chose not to recover from a recoverable error.)
+ - `"device-lost"`: things went horribly wrong, and the `WebGPUDevice` cannot be used anymore.
+   (An application may request a new device.)
```

Comment:
Nit: Might be a good idea to give examples of things that cause device lost and what things to do not. Obviously, unplugging the GPU, causing a TDR or disabling your display adapter in the control panel are examples of "going horribly wrong".  But I presume that validation failures or using invalid objects do not?  If so, seems a bit weird that only out-of-memory is referred to as 'recoverable' throughout the document.